zipline/pipeline/engine.py

```diff
@@ -273,6 +274,25 @@ def _inputs_for_term(self, term, workspace, graph):
             out.append(input_data)
         return out
 
+    def _similar_atomic_terms(self, graph, atomic_term):
```

Comment:
Since this is really about grouping terms by dispatched loader and extra_terms, I think it might be cleaner (also slightly faster?) to do this as a groupby at the top of `compute_chunk`.

```
from toolz import groupby, juxt # maps tuple of functions over a sequence of scalars

atomic_groups = groupby(juxt(dispatcher, graph.extra_rows.__getitem__), graph.atomic_terms)
```